src/session_impl.cpp

```diff
@@ -1807,16 +1805,68 @@ namespace aux {
 					// (which must be of the same family as the address we're
 					// connecting to)
 					if (device != ifs[k].name) continue;
+					eps.emplace_back(ifs[k].interface_address, port, device, ssl);
+				}
+			}
+		}
 
-					listen_socket_t const s = setup_listener(device
-						, tcp::endpoint(ifs[k].interface_address, port)
-						, flags | (ssl ? open_ssl_socket : 0), ec);
+		int const port_retries = m_settings.get_int(settings_pack::max_retry_port_bind);
 
-					if (!ec && s.sock)
-					{
-						m_listen_sockets.push_back(s);
-					}
-				}
+		// sockets we are keeping get moved to this list to prevent a socket from matching
+		// multiple endpoints
+		std::list<listen_socket_t> keep;
+
+		// remove any sockets which are no longer in the set of endpoints
+		// to listen on
+		// warning: O(n^2) operation!
+		// hopefully the system doesn't have too many interfaces
+		for (auto sock = m_listen_sockets.begin()
+			; sock != m_listen_sockets.end();)
+		{
+			auto match = std::find_if(eps.begin(), eps.end()
+				, [sock, port_retries](listen_endpoint_t const& ep)
+					{ return ep.ssl == sock->ssl
+						&& (ep.port == 0 || (sock->local_endpoint.port() >= ep.port
+							&& sock->local_endpoint.port() - ep.port < port_retries))
+						&& ep.device == sock->device
+						&& ep.addr == sock->local_endpoint.address(); });
+
+			if (match != eps.end())
+			{
+				// we don't need to create a new listen socket for this endpoint
+				// so remove it from the list
+				eps.erase(match);
+				keep.splice(keep.end(), m_listen_sockets, sock++);
+				continue;
+			}
+
+			// this socket's local_endpoint is not on the list of endpoints to listen on
+			// it has got to go
+			// TODO notify interested parties of this socket's demise
+#ifndef TORRENT_DISABLE_LOGGING
+			session_log("Closing listen socket for %s on device \"%s\""
+				, print_endpoint(sock->local_endpoint).c_str(), sock->device.c_str());
+#endif
+			if (sock->sock) sock->sock->close(ec);
+			if (sock->udp_sock) sock->udp_sock->close();
+			sock = m_listen_sockets.erase(sock);
```

Comment:
Currently handlers store a weak_ptr to the socket/acceptor so that they can detect if it has been deleted. If it turns out to be necessary to store a reference to listen_socket_t in a handler then my plan is to store listen_socket_t itself in a shared_ptr rather than the socket and acceptor.